### PR TITLE
[Issue Tracker] Watching Checkbox

### PR DESCRIPTION
**Pull request by AnyhowStep**
Watching (and un-watching) issues from the issue tracker, at the moment, is slow. You have to click on an issue to edit it, select "Yes" under "Watching?" and "Update" the issue.

You also cannot, at the moment, see which issues you are watching at a glance.

This pull request has a simple "Watching" column of the "All Active Issues" tab. The cells have a checkbox to indicate if you are currently watching it. Checking and unchecking them correspond to watching and un-watching them without going through the whole issue edit process.

`NDB_Menu_Filter.class.inc` has an extra member field called `$extra_query_params` now. This is necessary for having query params that are not fed by the user but by the code base, internally.

### CODE REVIEW

**Comment by kongtiaowang** [comment]
@AnyhowStep 
Using phpcbf to fix the format.

File: modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc [hunk 1]
```diff
@@ -41,19 +41,38 @@ class NDB_Menu_Filter_Form_Issue_Tracker extends NDB_Menu_Filter_Form
         $user =& User::singleton();
         $db   = Database::singleton();
 
+        //Why is this not a multi-line string?
```
Comment (alisterdev) [inline modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc]: You can leave these questions in the PR description and bring them up for discussion instead of polluting the code

File: php/installer/Installer.class.inc [hunk 2]
```diff
@@ -248,9 +248,6 @@ class Installer
         if ($this->_sourceSchemaFile($DB, "04-Help.sql") == false) {
             return false;
```
Comment (alisterdev) [inline php/installer/Installer.class.inc]: This doesn't seem to be a part of the PR, make sure to branch out of a fresh 17.1 branch

File: modules/issue_tracker/jsx/columnFormatter.js [hunk 3]
```diff
@@ -95,6 +133,9 @@ function formatColumn(column, cell, rowData, rowHeaders) {
       </td>
     );
   }
+  if (column === 'Watching') {
+    return (<WatchingCheckbox checked={cell==="1"} issue_id={row['Issue ID']}/>);
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/columnFormatter.js]: You should wrap the checkbox inside `<td>` here instead of directly inside the component. 
This is more explicit and it will make it easier to get rid of these `<td>`s in the future

File: modules/issue_tracker/jsx/columnFormatter.js [hunk 4]
```diff
@@ -1,5 +1,43 @@
...
+    }
+    render () {
+        return (
+            <td>
+                <input type="checkbox" checked={this.state.checked} onChange={this.onChange}/>
+            </td>
+        );
+    }
+    onChange (e) {
+        console.log(e._targetInst);
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/columnFormatter.js]: Leaving console logs in the code will definitely fail travis

File: modules/issue_tracker/jsx/columnFormatter.js [hunk 5]
```diff
@@ -1,5 +1,43 @@
 /* exported formatColumn */
 
+class WatchingCheckbox extends React.Component {
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/columnFormatter.js]: If a checkbox is a part of requirement, you could make it more generic and add as a part of `Form.js` form components. Otherwise, almost everything in Loris uses a dropdown instead. `<SelectElement>` component is  available for you in this case.

Comment (alisterdev) [inline modules/issue_tracker/jsx/columnFormatter.js]: Also, a new component should definitely not go inside `columnFormatter.js`

File: php/libraries/NDB_Menu_Filter.class.inc [hunk 6]
```diff
@@ -67,6 +67,15 @@ class NDB_Menu_Filter extends NDB_Menu
     var $filter;
 
     /**
+     * Extra query params for internal pselect() call
+     * Default null.
```
Comment (alisterdev) [inline php/libraries/NDB_Menu_Filter.class.inc]: These changes also seem irrelevant to the PR

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        ");
+        $db->execute($stmt, array(
+            "username"=>$user->getUsername(),
+            "issue_id"=>$issue_id
+        ));
+        echo json_encode(array(
+            "watching"=>false
+        ));
+    } else {
+        //Try to watch?
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: oh no

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        //I just realized the issues_watching table doesn't have an FK on the issueID column... wtf?
+        //Oh, well.
+        if ($db->pselectOne("
+            SELECT
+                EXISTS (
+                    SELECT
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: No sure what's up with this indentation

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: Indentation with multiline strings shouldn't be a thing Travis should look out for. Particularly, this relates to MySQL query formatting.

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: And how else would you format it?
````
SELECT
    <some value would need to be indented one level> (
        SELECT <subqueries should be indented one level>
    )
````

It's all for readability.

**Comment by alisterdev** [comment]
This definitely needs to be properly rebased from `17.1-dev`. There is a ton of irrelevant files/commits

**Comment by AnyhowStep** [comment]
Yikes. It was fine this morning.

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Issue tracker
+ *
+ * Toggles the issue-watching flag for a user.
+ *
+ * PHP Version 5
+ *
+ * @category Loris
+ * @package  Issue_Tracker
+ * @author   Caitrin Armstrong <caitrin.mcin@gmail.com>
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: @AnyhowStep You can put your name here, since you wrote the whole script, or put a generic `Loris Team <loris.mni@bic.mni.mcgill.ca>`

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: Same as below

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        array("error" => "Please log in")
+    );
+}
+
+if ($db->pselectOne(
+    "
+        SELECT
+            EXISTS (
+                SELECT
+                    *
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: I know @driusan doesn't like `select *`

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+         "issue_id" => $issue_id,
+        )
+    );
+    echo json_encode(
+        array("watching" => false)
+    );
+} else {
+    if ($db->pselectOne(
+        "
+            SELECT
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: I don't know if that much indentation makes it any more readable than SQL one-liners

File: modules/issue_tracker/jsx/WatchingCheckbox.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  }
+  render() {
+    return (
+      <td>
+        <input type = "checkbox" checked = {this.state.checked} onChange = {this.onChange}/>
+      </td>
+    );
+  }
+  onChange(e) {
+    var that = this;
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/WatchingCheckbox.js]: 1. Use `const` or `let` instead of `var` in all ES6 supported classes (a.k.a in all files that run through webpack)
2. Instead of using `self` or `that`, just bind `this` to success callback. e.g:
```js
success: function(data) {
  this.setState({checked: data.watching});
}.bind(this)
```

File: modules/issue_tracker/jsx/WatchingCheckbox.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    $.ajax({
+      type: "PUT",
+      dataType: "json",
+      url: "/issue_tracker/ajax/ToggleMyWatching.php?issue_id=" + this.state.issueId,
+      success: function(data) {
+        that.setState({
+          checked: data.watching
+        });
+      },
+      error: function(err) {}
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/WatchingCheckbox.js]: Probably want to use `console.error` to actually print the error to the console

File: modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc [hunk 15]
```diff
@@ -48,13 +48,29 @@ class NDB_Menu_Filter_Form_Issue_Tracker extends NDB_Menu_Filter_Form
...
         if (!$user->hasPermission('access_all_profiles')) {
             $site_arr     = implode(",", $user->getCenterID());
             $this->query .= " AND i.CenterID IN (" . $site_arr . ")";
         }
 
+        $this->extra_query_params = array(
+                                     "my_username" => $user->getUsername(),
+                                    );
+
+        $this->extra_query_params = array(
```
Comment (alisterdev) [inline modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc]: Next level code duplication

Comment (AnyhowStep) [inline modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc]: Silly me

Comment (AnyhowStep) [inline modules/issue_tracker/php/NDB_Menu_Filter_Form_issue_tracker.class.inc]: Must've merged wrongly or something

File: modules/issue_tracker/ajax/ToggleMyWatching.php (hunk 11 above)
Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: `SELECT *` for `EXISTS` is perfectly fine.

File: modules/issue_tracker/ajax/ToggleMyWatching.php (hunk 12 above)
Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: It's not my fault. It's PHPCBF

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: The only SQL one-liner I'll do is `SELECT 1;`

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: I'm looking at my file now and it's honestly phpcbf and phpcs enforcing these indentation standards. If it were up to me, the strings wouldn't look like that.

File: modules/issue_tracker/ajax/ToggleMyWatching.php (hunk 9 above)
Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: Thanks for the convenient copy pasta!

File: modules/issue_tracker/ajax/ToggleMyWatching.php (hunk 12 above)
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: By one liners I mean writing that stuff as
`SELECT * FROM issues WHERE issueID = :issue_id` which phpcs doesn't have a problem with

File: modules/issue_tracker/ajax/ToggleMyWatching.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+                        *
+                    FROM
+                        issues
+                    WHERE
+                        issueID = :issue_id
+                )
+        ",
+        array("issue_id" => $issue_id)
+    )) {
+        $stmt = $db->prepare(
```
Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: @driusan pointed out that there's a `$db` function that `prepare` and `execute` for you

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: Which?

Comment (alisterdev) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: I tagged him to confirm, but it seems that `$db->delete()` does what you need

Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: Found corresponding `$db->insert()`, too. Thanks.

File: modules/issue_tracker/ajax/ToggleMyWatching.php (hunk 12 above)
Comment (AnyhowStep) [inline modules/issue_tracker/ajax/ToggleMyWatching.php]: phpcs has a problem with multiline strings. Not multiline sql queries.
It would still attempt to change it to:
```
some_function_call(
    "
        Some ugly indented
        multiline string.
    "
);
```

At least it leaves the formatting of the SQL query *inside* the string alone. All I ask is for readable SQL statements.

**Comment by AnyhowStep** [comment]
Rebased again

File: modules/issue_tracker/jsx/WatchingCheckbox.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    super();
+    this.state = {
+      checked: prop.checked,
+      issueId: prop.issue_id
+    };
+    this.onChange = this.onChange.bind(this);
+  }
+  render() {
+    return (
+      <td>
```
Comment (alisterdev) [inline modules/issue_tracker/jsx/WatchingCheckbox.js]: Remove this `<td>`

image

**Comment by AnyhowStep** [comment]
Whoops. Fixed

**Comment by alisterdev** [comment]
Travis fails on a non issue tracker related class, so it feels like this PR needs a rebase

**Comment by jstirling91** [comment]
@AnyhowStep Please rebase and fix conflicts

File: modules/issue_tracker/jsx/WatchingCheckbox.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    super();
+    this.state = {
+      checked: prop.checked,
+      issueId: prop.issue_id
+    };
+    this.onChange = this.onChange.bind(this);
+  }
+  render() {
+    return (
+      <input type = "checkbox" checked = {this.state.checked} onChange = {this.onChange}/>
```
Comment (jstirling91) [inline modules/issue_tracker/jsx/WatchingCheckbox.js]: You should just use props here instead of states

File: modules/issue_tracker/jsx/WatchingCheckbox.js [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  onChange(e) {
+    $.ajax({
+      type: "PUT",
+      dataType: "json",
+      url: "/issue_tracker/ajax/ToggleMyWatching.php?issue_id=" + this.state.issueId,
+      success: function(data) {
+        this.setState({
+          checked: data.watching
+        });
```
Comment (jstirling91) [inline modules/issue_tracker/jsx/WatchingCheckbox.js]: Should pass a function as a prop to update this in the parent class

**Comment by xlecours** [comment]
The requested changes wont be done.
